design/002-secret-generators.md

```diff
@@ -0,0 +1,90 @@
+```yaml
+---
+title: Secret Generators
+version: v1alpha1
+authors: Christian Hünig, Jan Steffen
+creation-date: 2022-07-08
+status: draft
+---
+```
+
+# Secret Generators
+
+## Table of Contents
+
+<!-- toc -->
+// autogen please
+<!-- /toc -->
+
+## Summary
+
+There are use-cases where a secret could be generated by the operator instead of using pre-existing ones from the secret store. For example a MongoDB could be spun up using a password generated by the operator. This password is then used by an application in the cluster to access the database. No human ever needs to read that secret or have to specify it beforehand. A similar feature is available in a project called [VaultOperator](https://github.com/finleap-connect/vaultoperator/) which tries to solve similar problems like this operator just for Hashicorp Vault as a secret store.
+
+## Motivation
+
+The motivation is to boost adoption of zero-trust secrets management
+and to make it simple to create secure secrets for machine users
+of services running inside clusters (i.e. databases etc).
+
+### Goals
+
+* enable zero-trust secret management for purely technical and internal users
+* enable secure passwords by default
+* reduce error rates when creating RSA or ECDSA private keys
+
+### Non-Goals
+
+* auto-rotation of secrets
+
+## Proposal
+
+Implement several generators to support a number of data types.
+A generator should be abstracted via an interface and thus be extendable for future types.
+
+To begin with we should support the following generators:
+
+* `string` generates a random string with length `args[0]`
+* `bytes` generates random bytes with length `args[0]`
+* `password` special form of string generations where `args[0]` is the length and is mandatory. `args[1]` optionally specifies the number of digits and `args[2]` optionally defines the number of symbols.
+* `rsa` generates RSA private key with bit size `args[0]` (encoded as PEM)
+* `ecdsa` generates EC private key with curve `args[0]` (encoded as PEM)
+
+### API
+
+```yaml
+apiVersion: external-secrets.io/v1beta1
+kind: ExternalSecret
+metadata:
+  name: example
+spec:
+  refreshInterval: 1h
+  secretStoreRef:
+    name: secretstore-sample
+    kind: SecretStore
+  target:
+    name: secret-to-be-created
+    creationPolicy: Owner
+  data:
+  - secretKey: secret-key-to-be-managed
+    remoteRef:
+      key: provider-key
+      version: provider-key-version
+      property: provider-key-property
+    generator: # optional, will create a random 16 char string if key not present
+      name: "string"
+      args: [16]
```

Comment:
This should be strongly typed, provide an `apiVersion` to support versioning and must be extensible so we are able to add other generators.